rfcs/0004-fulltext-search.md

```diff
@@ -0,0 +1,275 @@
+for end-use cases. 
+
+For each fulltext search index a subgraph developer must be able to specify:
+    1. a language, 
+    2. a set of text document fields to include,
+    3. relative weighting for each field in the index.
+    4. a choice of ranking algorithm for sorting query result items
+
+The proposed process of adding one or more fulltext search index to an Entity 
+involves adding one or more fulltext directive to the `_Schema_` type in the 
+subgraph's GraphQL schema.  Each fulltext definition will have four required
+top level parameters: `name`, `language`, `algorithm`, and `include`. 
+The fulltext search definitions will be used to generate queries on the GraphQL
+schema that will be exposed to the end user.  
+
+Applying fulltext search indexes across entities will be a powerful 
+abstraction allowing users to search across all relevant entities in one
+query. An index search across entities will by definition have polymorphic
+results, so in that case a union type will also be generated in the schema 
+for the query result items. 
+
+Verification of the index definition will ensure that all fields referenced 
+are valid String type fields. To be clear, the proposed interface will add 
+the fulltext search index definitions to the subgraph deployment hash, so 
+adding or updating a fulltext search index will require an update to the subgraph. 
+With subgraph composition it will be possible to easily create new subgraphs 
+that add specific fulltext search indexes to an existing subgraph. 
+
+Example fulltext search definition:
+
+```graphql
+type _Schema_ 
+  @fulltext(
+    name: "media"
+    ...
+  )
+  @fulltext(
+    name: "search",
+    language: "english",
+    algorithm: "ranked",
+    include: [
+      {
+        entity: "Band",
+        fields: [
+          { name: "name", weight: 5 },
+        ]
+      },
+      {
+        entity: "Album",
+        fields: [
+          { name: "title", weight: 5 },
+        ]
+      },
+      {
+        entity: "Musician",
+        fields: [
+          { name: "name", weight: 10 },
+          { name: "bio", weight: 5 },
+        ]
+      }
+    ]
+  )
+```
+
+The schema generated from the above definition: 
+```graphql
+union FulltextMediaResultItem = ...
+union FulltextSearchResultItem = Band | Album | Musician
+type Query {
+  media...
+  search(text: String!, first: Int, skip: Int, block: Int): [FulltextSearchResultItem!]!
+}
+```
+
+### GraphQL Query interface
+
+End users of the subgraph will have access to the queries generated for the
+fulltext search indexes alongside the other queries generated for each 
+entity in the subgraph.  In the case of a fulltext search index that spans
+multiple entities, [inline fragments](https://graphql.org/learn/queries/#inline-fragments) 
+may be used in the query to deal with the polymorphic result items. In the 
+front-end the `__typename` in each result item can be used to render the 
+components accordingly.  
```

Comment:
Which components? We could just say `the `__typename` field can be used to distinguish the concrete entity types of the returned results`.